File: src/main/java/com/ject/studytrip/mission/infra/querydsl/DailyMissionQueryRepositoryAdapter.java
package com.ject.studytrip.mission.infra.querydsl;

import static com.ject.studytrip.mission.domain.model.QDailyMission.dailyMission;
import static com.ject.studytrip.mission.domain.model.QMission.mission;
import static com.ject.studytrip.stamp.domain.model.QStamp.stamp;

import com.ject.studytrip.mission.domain.model.DailyMission;
import com.ject.studytrip.mission.domain.repository.DailyMissionQueryRepository;
import com.querydsl.jpa.impl.JPAQueryFactory;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Repository;

@Repository
@RequiredArgsConstructor
public class DailyMissionQueryRepositoryAdapter implements DailyMissionQueryRepository {
    private final JPAQueryFactory queryFactory;

    @Override
    public List<DailyMission> findAllByDailyGoalIdFetchJoinMission(Long dailyGoalId) {
        return queryFactory
                .selectFrom(dailyMission)
                .join(dailyMission.mission, mission)
                .fetchJoin()
                .where(dailyMission.dailyGoal.id.eq(dailyGoalId), dailyMission.deletedAt.isNull())
                .fetch();
    }

    @Override
    public List<DailyMission> findAllWithMissionAndStampByIds(List<Long> ids) {
        return queryFactory
                .selectFrom(dailyMission)
                .join(dailyMission.mission, mission)
                .fetchJoin()
                .join(mission.stamp, stamp)
                .fetchJoin()
                .where(dailyMission.id.in(ids))
                .fetch();
    }
}

File: src/main/kotlin/com/ject/studytrip/mission/infra/querydsl/DailyMissionQueryRepositoryAdapter.kt
package com.ject.studytrip.mission.infra.querydsl

import com.ject.studytrip.mission.domain.model.DailyMission
import com.ject.studytrip.mission.domain.model.QDailyMission.dailyMission
import com.ject.studytrip.mission.domain.model.QMission.mission
import com.ject.studytrip.mission.domain.repository.DailyMissionQueryRepository
import com.ject.studytrip.stamp.domain.model.QStamp.stamp
import com.querydsl.jpa.impl.JPAQueryFactory
import org.springframework.stereotype.Repository

@Repository
class DailyMissionQueryRepositoryAdapter(
    private val queryFactory: JPAQueryFactory,
) : DailyMissionQueryRepository {
    override fun findAllByDailyGoalIdFetchJoinMission(dailyGoalId: Long): List<DailyMission> =
        queryFactory
            .selectFrom(dailyMission)
            .join(dailyMission.mission, mission)
            .fetchJoin()
            .where(
                dailyMission.dailyGoal.id.eq(dailyGoalId),
                dailyMission.deletedAt.isNull,
            ).fetch()

    override fun findAllWithMissionAndStampByIds(ids: List<Long>): List<DailyMission> =
        queryFactory
            .selectFrom(dailyMission)
            .join(dailyMission.mission, mission)
            .fetchJoin()
            .join(mission.stamp, stamp)
            .fetchJoin()
            .where(dailyMission.id.`in`(ids))
            .fetch()
}
